src/solver/src/steady_mesh_adaptive_solver.C

```diff
@@ -93,71 +95,93 @@ namespace GRINS
 
     for ( unsigned int r_step = 0; r_step < _mesh_adaptivity_options.max_refinement_steps(); r_step++ )
       {
-        std::cout << "==========================================================" << std::endl
+        std::cout << std::endl
+                  << "==========================================================" << std::endl
                   << "Adaptive Refinement Step " << r_step << std::endl
                   << "==========================================================" << std::endl;
 
-        // Solve the forward problem
-        context.system->solve();
+        converged = this->amr_helper(context,r_step,true);
+        if (converged)
+          break;
 
-        if( context.output_vis )
-          {
-            context.postprocessing->update_quantities( *(context.equation_system) );
-            context.vis->output( context.equation_system );
-          }
+      } // r_step for-loop
 
-        // Solve adjoint system
-        if(context.do_adjoint_solve)
-          this->steady_adjoint_solve(context);
+    // this will print output the error estimates and/or meshes after the final AMR run
+    if (_mesh_adaptivity_options.max_refinement_steps() > 0)
+      if (! converged)
+        {
+          std::cout << std::endl
+                    << "==========================================================" << std::endl
+                    << "Post-AMR Output " << std::endl
+                    << "==========================================================" << std::endl;
+          this->amr_helper(context,_mesh_adaptivity_options.max_refinement_steps(),false);
+        }
 
-        if(context.output_adjoint)
-          context.vis->output_adjoint(context.equation_system, context.system);
+    return;
+  }
 
-        if( context.output_residual )
-          {
-            context.vis->output_residual( context.equation_system, context.system );
-          }
+  bool SteadyMeshAdaptiveSolver::amr_helper(SolverContext & context, unsigned int r_step, bool do_amr)
+  {
+    // Solve the forward problem
+    context.system->solve();
 
-        // Now we construct the data structures for the mesh refinement process
-        libMesh::ErrorVector error;
-        this->estimate_error_for_amr( context, error );
+    if( context.output_vis )
+      {
+        context.postprocessing->update_quantities( *(context.equation_system) );
 
-        // Get the global error estimate if you can and are asked to
-        if( _error_estimator_options.compute_qoi_error_estimate() )
-          for(unsigned int i = 0; i != context.system->qoi.size(); i++)
-            {
-              libMesh::AdjointRefinementEstimator* adjoint_ref_error_estimator = libMesh::cast_ptr<libMesh::AdjointRefinementEstimator*>( context.error_estimator.get() );
-              std::cout<<"The error estimate for QoI("<<i<<") is: "<<adjoint_ref_error_estimator->get_global_QoI_error_estimate(i)<<std::endl;
-            }
+        if (context.output_every_amr)
```

Comment:
Instead of a bool, maybe an integer to output every n refinement steps?

Reply:
I guess that would make sense, especially in the context of the `UnsteadyMeshAdaptiveSolver`

Reply:
Yeah, I was actually thinking about unsteady case, but yeah, for steady case, let's be able to reuse that code and if we're just and if the user put something > 1, we can flash a warning and switch it back to 1.

Reply:
Sorry, not quite sure what you're saying here...

Reply:
What my stream-of-consciousness comments don't make sense?! :P Let's use integer for steady case, but if the user puts anything other than 0 or 1, we print a warning and change it back to 0 or 1 (whatever makes sense).

Reply:
Hadn't had my morning tea yet, that makes more sense haha